src/py21cmfast/drivers/param_config.py

```diff
@@ -0,0 +1,440 @@
+                warnings.warn(msg)
+            else:
+                raise ValueError(msg)
+
+    def keys(self):
+        """The list of available structs in this instance."""
+        # Allow using **input_parameters
+        return [
+            field.name
+            for field in attrs.fields(self.__class__)
+            if getattr(self, field.name) is not None
+        ]
+
+    def __getitem__(self, key):
+        """Get an item from the instance in a dict-like manner."""
+        # Also allow using **input_parameters
+        return getattr(self, key)
+
+    def is_compatible_with(self, other: InputParameters) -> bool:
+        """Check if this object is compatible with another parameter struct.
+
+        Compatibility is slightly different from strict equality. Compatibility requires
+        that if a parameter struct *exists* on the other object, then it must be equal
+        to this one. That is, if astro_params is None on the other InputParameter object,
+        then this one can have astro_params as NOT None, and it will still be
+        compatible. However the inverse is not true -- if this one has astro_params as
+        None, then all others must have it as None as well.
+        """
+        if not isinstance(other, InputParameters):
+            return False
+
+        return not any(
+            other[key] is not None and self[key] != other[key] for key in self.keys()
+        )
+
+    def merge(self, other: InputParameters) -> InputParameters:
+        """Merge another InputParameters instance with this one, checking for compatibility."""
+        if not self.is_compatible_with(other):
+            raise ValueError("Input parameters are not compatible.")
+
+        return attrs.evolve(self, **other)
+
+    @classmethod
+    def combine(cls, inputs: Sequence[InputParameters]) -> InputParameters:
+        """Combine multiple input parameter structs into one.
+
+        Parameters
+        ----------
+        inputs : list of :class:`InputParameters`
+            The input parameter structs to combine.
+
+        Returns
+        -------
+        :class:`InputParameters`
+            The combined input parameter struct.
+        """
+        # Create an empty instance of InputParameters
+        this = inputs[0]
+
+        for inp in inputs[1:]:
+            this = this.merge(inp)
+
+        return this
+
+    @classmethod
+    def from_output_structs(cls, output_structs, **defaults) -> InputParameters:
+        """Generate a new InputParameters instance given a list of OutputStructs."""
+        input_params = []
+        for struct in output_structs:
+            if struct is not None:
+                for k, v in dir(struct).items():
+                    if isinstance(v, InputParameters):
+                        input_params.append(v)
+
+        if len(input_params) == 0:
+            return cls(**defaults)
+        else:
+            out = cls.combine(input_params)
+            return out.merge(cls(**defaults))
```

Comment:
Looks like `**defaults` are prioritised over the inputs (in `merge()`) BUT it doesn't matter because they're always forced to be consistent with one another. We should put that in the docstring though.

Reply:
Some issues have been addressed here where default user/cosmo params were being merged when we only gave astro/flag params. I might need some feedback though since I implemented it by allowing partially filled `InputParameters` structs. They should still throw errors when trying to build a struct which needs a certain parameter struct but we no longer set defaults in the `InputParameters`.

Reply:
OK, I will have a look.